google/resource_storage_object_acl.go

```diff
@@ -56,196 +102,247 @@ func resourceStorageObjectAclCreate(d *schema.ResourceData, meta interface{}) er
+// Creates 3 lists of changes we need to make to go from one set of entities to another- which entities need to be created, update, and deleted
+// Not resource specific
+func getRoleEntityChange(old []interface{}, new []interface{}, owner string) ([]*RoleEntity, []*RoleEntity, []*RoleEntity, error) {
+	newEntitiesUsed := make(map[string]struct{})
+	for _, v := range new {
+		res, err := getRoleEntityPair(v.(string))
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("New state has malformed Role/Entity pair: %v", err)
+		}
 
-			if err != nil {
-				return fmt.Errorf("Error updating ACL for object %s: %v", bucket, err)
-			}
+		if _, ok := newEntitiesUsed[res.Entity]; ok {
+			return nil, nil, nil, fmt.Errorf("New state has duplicate Entity: %v", res.Entity)
 		}
 
-		for entity, _ := range old_re_map {
-			log.Printf("[DEBUG]: removing entity %s", entity)
-			err := config.clientStorage.ObjectAccessControls.Delete(bucket, object, entity).Do()
+		newEntitiesUsed[res.Entity] = struct{}{}
+	}
 
-			if err != nil {
-				return fmt.Errorf("Error updating ACL for object %s: %v", bucket, err)
-			}
+	oldEntitiesUsed := make(map[string]string)
+	for _, v := range old {
+		res, err := getRoleEntityPair(v.(string))
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("Old state has malformed Role/Entity pair: %v", err)
 		}
 
-		return resourceStorageObjectAclRead(d, meta)
+		// Updating the owner will error out, so let's avoid it.
+		if res.Entity == owner {
+			continue
+		}
+
+		oldEntitiesUsed[res.Entity] = res.Role
 	}
 
-	return nil
-}
+	var create []*RoleEntity
+	var update []*RoleEntity
+	var remove []*RoleEntity
+	for _, re := range new {
+		res, err := getRoleEntityPair(re.(string))
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("New state has malformed Role/Entity pair: %v", err)
+		}
 
-func resourceStorageObjectAclDelete(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*Config)
+		// Updating the owner will error out, so let's never do it.
+		if res.Entity == owner {
+			continue
+		}
 
-	bucket := d.Get("bucket").(string)
-	object := d.Get("object").(string)
+		if v, ok := oldEntitiesUsed[res.Entity]; ok {
+			if res.Role != v {
+				update = append(update, res)
+			}
 
-	re_local := d.Get("role_entity").([]interface{})
-	for _, v := range re_local {
+			delete(oldEntitiesUsed, res.Entity)
+		} else {
+			create = append(create, res)
+		}
+	}
+
+	for _, v := range old {
 		res, err := getRoleEntityPair(v.(string))
 		if err != nil {
-			return err
+			return nil, nil, nil, fmt.Errorf("Old state has malformed Role/Entity pair: %v", err)
 		}
 
-		entity := res.Entity
+		if _, ok := oldEntitiesUsed[res.Entity]; ok {
+			remove = append(remove, res)
+		}
+	}
 
-		log.Printf("[DEBUG]: removing entity %s", entity)
+	return create, update, remove, nil
+}
+
+// Takes in 3 lists of changes for a Storage Object and makes those changes, ignoring the bucket owner entity - they can't be modified
```

Comment:
technically this function doesn't ignore the bucket owner entity- by the time it reaches here, it's already been ignored.